polars/polars-lazy/polars-plan/src/dsl/function_expr/temporal.rs

```diff
@@ -7,43 +7,71 @@ use polars_time::prelude::*;
 use super::*;
 
 #[cfg(feature = "date_offset")]
-pub(super) fn date_offset(s: Series, offset: Duration) -> PolarsResult<Series> {
+pub(super) fn date_offset(s: &[Series]) -> PolarsResult<Series> {
+    let sa = &s[0];
+    let sb = &s[1];
+
     let preserve_sortedness: bool;
-    let out = match s.dtype().clone() {
+    let out = match sa.dtype().clone() {
         DataType::Date => {
-            let s = s
+            let sa = sa
                 .cast(&DataType::Datetime(TimeUnit::Milliseconds, None))
                 .unwrap();
             preserve_sortedness = true;
-            date_offset(s, offset).and_then(|s| s.cast(&DataType::Date))
+            date_offset(&[sa, sb.clone()]).and_then(|s| s.cast(&DataType::Date))
         }
         DataType::Datetime(tu, tz) => {
-            let ca = s.datetime().unwrap();
+            let ca = sa.datetime().unwrap();
+            let offsets = sb.utf8().unwrap();
 
-            fn offset_fn(tu: TimeUnit) -> fn(&Duration, i64, Option<&Tz>) -> PolarsResult<i64> {
-                match tu {
-                    TimeUnit::Nanoseconds => Duration::add_ns,
-                    TimeUnit::Microseconds => Duration::add_us,
-                    TimeUnit::Milliseconds => Duration::add_ms,
-                }
-            }
+            let offset_fn = match tu {
+                TimeUnit::Nanoseconds => Duration::add_ns,
+                TimeUnit::Microseconds => Duration::add_us,
+                TimeUnit::Milliseconds => Duration::add_ms,
+            };
 
-            let out = match tz {
+            let tz_args = match tz {
                 #[cfg(feature = "timezones")]
-                Some(ref tz) => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, tz.parse::<Tz>().ok().as_ref()))
+                Some(ref tz) => tz.parse::<Tz>().ok(),
+                _ => None,
+            };
+
+            let out = match offsets.len() {
+                1 => {
+                    let offset = match offsets.get(0) {
+                        Some(offset) => Duration::parse(offset),
+                        _ => Duration::new(0),
+                    };
+                    ca.0.try_apply(|v| offset_fn(&offset, v, tz_args.as_ref()))
                 }
                 _ => {
-                    let offset_fn = offset_fn(tu);
-                    ca.0.try_apply(|v| offset_fn(&offset, v, None))
+                    let out = ca
```

Comment:
This logic must be implemented in `polars-ops` and be implemented on a `PrimitiveArray` as per the example I showed.

Iterating over `ChunkedArrays` is slower than iterating over `PrimtiveArrays`

Reply:
Yeah, thanks @ritchie46 - I did try to implement that using the example template, https://github.com/pola-rs/polars/pull/9967#discussion_r1271059931

In this case we have:

`s_1` of type `datetime`

`s_2` of type `str`

With `.downcast_iter()` I end up with `s_1` being `PrimitiveArray` and `s_2` being `Utf8Array`

So I was a bit unsure how/where one is supposed to run the strings through `Duration::parse()` and whether a Series of type `duration` will actually cast into a `PrimitiveArray`.

It seems I may have underestimated what was needed to add this particular feature - it's probably counterproductive for me to be involved and requires someone with proper Rust knowledge/experience.

Reply:
Ok, I can pick it up later. Can you leave this PR open? Thanks!

Reply:
Sure no problem.

Reply:
I've tried taking this forwards